kernel/src/scan/mod.rs

```diff
@@ -94,23 +98,171 @@ impl ScanBuilder {
+        let state_info = get_state_info(
             logical_schema.as_ref(),
             &self.snapshot.metadata().partition_columns,
         )?;
-        let physical_schema = Arc::new(StructType::new(read_fields));
+
+        let physical_predicate = match self.predicate {
+            Some(predicate) => PhysicalPredicate::try_new(&predicate, &logical_schema)?,
+            None => PhysicalPredicate::None,
+        };
 
         Ok(Scan {
             snapshot: self.snapshot,
             logical_schema,
-            physical_schema,
-            predicate: self.predicate,
-            all_fields,
-            have_partition_cols,
+            physical_schema: Arc::new(StructType::new(state_info.read_fields)),
+            physical_predicate,
+            all_fields: state_info.all_fields,
+            have_partition_cols: state_info.have_partition_cols,
         })
     }
 }
 
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) enum PhysicalPredicate {
+    Some(ExpressionRef, SchemaRef),
+    StaticSkipAll,
+    None,
+}
+
+impl PhysicalPredicate {
+    /// If we have a predicate, verify the columns it references and apply column mapping. First, get
+    /// the set of references; use that to filter the schema to only the columns of interest (and
+    /// verify that all referenced columns exist); then use the resulting logical/physical mappings
+    /// to rewrite the expression with physical column names.
+    ///
+    /// NOTE: It is possible the predicate resolves to FALSE even ignoring column references,
+    /// e.g. `col > 10 AND FALSE`. Such predicates can statically skip the whole query.
+    pub(crate) fn try_new(
+        predicate: &Expression,
+        logical_schema: &Schema,
+    ) -> DeltaResult<PhysicalPredicate> {
+        if can_statically_skip_all_files(predicate) {
+            return Ok(PhysicalPredicate::StaticSkipAll);
+        }
+        let mut get_referenced_fields = GetReferencedFields {
+            unresolved_references: predicate.references(),
+            column_mappings: HashMap::new(),
+            logical_path: vec![],
+            physical_path: vec![],
+        };
+        let schema_opt = get_referenced_fields.transform_struct(logical_schema);
+        let mut unresolved = get_referenced_fields.unresolved_references.into_iter();
+        if let Some(unresolved) = unresolved.next() {
+            // Schema traversal failed to resolve at least one column referenced by the predicate.
+            //
+            // NOTE: It's a pretty serious engine bug if we got this far with a query whose WHERE
+            // clause has invalid column references. Data skipping is best-effort and the predicate
+            // anyway needs to be evaluated against every row of data -- which is impossible if the
+            // columns are missing/invalid. Just blow up instead of trying to handle it gracefully.
+            return Err(Error::missing_column(format!(
```

Comment:
One thing we'll want to watch out for in the future is a predicate like `where id = 2 and _change_type = "insert"`.

The `_change_type` part of the predicate can't be applied to the physical data, so it should be dropped.

Same thing applies to generated columns and partition columns I think?

Reply:
We're checking against the _logical_ schema, which includes partition columns and also (hopefully) includes the `_change_type` and other generated columns? 

Meanwhile, we have disabled the optimization in the parquet skipping code, that would treat missing physical columns as all-NULL. If we ever tried to re-enable that optimization, we would need to be very sure we've classified the columns correctly first.

Reply:
>  we have disabled the optimization in the parquet skipping code, that would treat missing physical columns as all-NULL

Aha I'd expected such an optimization to exist here in the physical predicate. Since we handle that elsewhere, then we're good.

> we would need to be very sure we've classified the columns correctly first

 If we had access to `Vec<ColumnType>`, we'd know how every column is classified.

Reply:
Yes, and eventually we'll need that to support data skipping over partition columns, since we do have the metadata for those but it's in a different column and requires slightly different logic.